libdnf-cli/progressbar/download_progress_bar.cpp

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright (C) 2020 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#include "libdnf-cli/progressbar/download_progress_bar.hpp"
+#include "libdnf-cli/utils/tty.hpp"
+
+#include <algorithm>
+#include <iomanip>
+
+
+using namespace libdnf::cli::utils;
+
+
+namespace libdnf::cli::progressbar {
+
+
+// Columns (widgets):
+// <numbers><description...................><%%%><progress............><speed.....><size.....><time....>
+
+// Example:
+// [  1/999] filename                                                 | 123.4 kB/s | 123.4 kB |  00m01s
+// [  2/999] filename                       100% [====================] 123.4 kB/s | 123.4 kB | -00m02s
+// ----------------------------------------------------------------------------------------------------
+// [  1/999] Total                           10% [===                 ] 543.2 kB/s | 543.2 MB | -12m34s
+
+
+DownloadProgressBar::DownloadProgressBar(int64_t download_size, const std::string & description)
+    : ProgressBar(download_size, description) {
+    number_widget.set_bar(this);
+    description_widget.set_bar(this);
+    percent_widget.set_bar(this);
+    progress_widget.set_bar(this);
+    speed_widget.set_bar(this);
+    size_widget.set_bar(this);
+    time_widget.set_bar(this);
+}
+
+
+static std::size_t get_bar_width(std::vector<Widget *> widgets) {
+    std::size_t result = 0;
+    for (Widget * widget : widgets) {
+        result += widget->get_total_width();
+    }
+    return result;
+}
+
+
+std::ostream & operator<<(std::ostream & stream, DownloadProgressBar & bar) {
+    bar.to_stream(stream);
+    return stream;
+}
+
+
+void DownloadProgressBar::to_stream(std::ostream & stream) {
+    update();
+
+    // set the default delimiters
+    number_widget.set_delimiter_before("");
```

Comment:
Do these setters really have to be called every time to_stream is called?
What about move them into `DownloadProgressBar` constructor?

Reply:
OK. I see now. `speed_widget.set_delimiter_before(" ");` may be called during computing widgets to show.

Reply:
Exactly, the delimiter changes when widgets get excluded from the bar to fit the terminal width.